keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,879 @@
+
+##### Risk
+
+OpenAPI v3 data is not available in the current cluster.
+
+##### Mitigation
+
+If the initial request for OpenAPI V3 fails, the old implementation using
+OpenAPI v2 should be used instead. There is another risk that this fallback process
+takes too long for interactive speed, so timeout should be carefully considered
+to prevent user from waiting too long to see their results. 
+
+#### Template API Stability
+##### Risk
+
+If templates are written to target a specific OpenAPI version, then changes to
+the format would case templates to be broken. It is a goal of this KEP to be more
+maintainable than the current verison. If we risk templates breaking due to format
+changes, that voids the goals of the KEP.
+
+This is only a risk of users are allowed to specify their own templates. If all
+templates are maintained internally, then API stability it is something to consider
+but not necessarily a requirement.
+
+##### Mitigation
+
+TBD
+
+#### OpenAPI serialization time
+##### Risk
+
+Today there is no interactive-speed way to deserialize protobuf or JSON openapi
+v3 data into the kube-openapi format.
+
+##### Mitigation
+
+There has been recent progress in this area. To unmarshal kube-OpenAPI v3 is now able
+to be done in a performant enough way to do it in the CLI. This KEP's beta release
+should be blocked on the merging of this optimization.
+
+## Design Details
+
+#### Current High-level Approach
+
+TODO: fill with links to referred implementation
+
+1. User types `kubectl explain pods`
+2. kubectl resolves 'pods' to GVR core v1 pods using cluster discovery information
+3. kubectl resolves GVR to its GVK using restmapper
+4. kubectl fetches `/openapi/v2` as protobuf
+5. kubectl parses the protobuf into `gnostic_v2.Document`
+6. kubectl converts `gnostic_v2.Document` into `proto.Models`
+7. kubectl searches the document's `Definitions` for a schema with the
+extension `x-kubernetes-group-version-kind` matching the interested GVK
+8. kubectl renders the definition using its hardcoded printer
+9. If `--recursive` was used, repeat step 8 for the transitive closure of
+  object-typed fields of the top-level object. Concat the results together.
+
+#### Proposed High-level Approach
+
+1. User types `kubectl explain pods`
+2. kubectl resolves 'pods' to GVR core v1 pods using cluster discovery information
+3. kubectl fetches `/openapi/v3/<group>/<version>`
+4. kubectl parses the result as kube-openapi spec3
+5. kubectl locates the return type of the `/apis/<group>/<version>/<resource>` OpenAPI Path
+6. kubectl renders the type using its built-in template for human-readable plaintext
+7. If `--recursive` was used, repeat step 6 for the transitive closure of
+  object-typed fields of the top-level object. Concat the results together.
+
+### Template rendering
+
+Go's text/template will be used due to its familiarity, stability, and virtue of being in stdlib.
+
+
+The following library functions will be provided to templates:
```

Comment:
It would be good to describe some of the principles under which we want to add library functions to templates, and the rules that apply (we don't remove functions, we version them, we don't add functions like Y or X, etc), to guide reviewers later.

Reply:
To focus the scope of this KEP I've removed the proposal to allow custom user templates. Once we have settled on best practices/good library functions in our proposed internal `md` `web`, and `raw` outputs, I think custom user templates is a feature that can be perhaps proposed in a future KEP after an additinonal use case manifests. 

There is also the argument that a lot of tooling can be built around the `raw` output so it may ultimately not be necessary to provide user custom templates.